packages/slate/src/models/node.js

```diff
@@ -1061,120 +1053,129 @@ class Node {
-
-    return this.getCharactersAtRange(range).reduce((memo, char) => {
-      if (char) {
-        char.marks.toArray().forEach(c => memo.push(c))
-      }
-      return memo
-    }, [])
+    if (range.isUnset) return OrderedSet()
+    if (range.isCollapsed) {
+      // PREF: range is not cachable, use key and offset as proxies for cache
+      return this.getMarksAtPosition(range.startKey, range.startOffset)
+    }
+    const { startKey, startOffset, endKey, endOffset } = range
+    return this.getOrderedMarksBetweenPositions(
+      startKey,
+      startOffset,
+      endKey,
+      endOffset
+    )
   }
 
   /**
-   * Get a set of the marks in a `range` for insertion behavior.
+   * Get a set of the marks in a `range`.
+   * PREF: arguments use key and offset for utilizing cache
    *
-   * @param {Range} range
-   * @return {Array}
+   * @param {string} startKey
+   * @param {number} startOffset
+   * @param {string} endKey
+   * @param {number} endOffset
+   * @returns {OrderedSet<Mark>}
    */
 
-  getInsertMarksAtRangeAsArray(range) {
-    range = range.normalize(this)
-    if (range.isUnset) return []
-    if (range.isCollapsed) return this.getMarksAtCollapsedRangeAsArray(range)
+  getOrderedMarksBetweenPositions(startKey, startOffset, endKey, endOffset) {
+    if (startKey === endKey) {
+      const startText = this.getDescendant(startKey)
+      return startText.getMarksBetweenOffsets(startOffset, endOffset)
+    }
 
-    const text = this.getDescendant(range.startKey)
-    const char = text.characters.get(range.startOffset)
-    if (!char) return []
+    const texts = this.getTextsBetweenPositionsAsArray(startKey, endKey)
 
-    return char.marks.toArray()
+    return OrderedSet().withMutations(result => {
+      texts.forEach(text => {
+        if (text.key === startKey) {
+          result.union(
+            text.getMarksBetweenOffsets(startOffset, text.text.length)
+          )
+        } else if (text.key === endKey) {
+          result.union(text.getMarksBetweenOffsets(0, endOffset))
+        } else {
+          result.union(text.getMarks())
+        }
+      })
+    })
   }
 
   /**
-   * Get a set of marks in a `range`, by treating it as collapsed.
+   * Get a set of the active marks in a `range`.
    *
    * @param {Range} range
-   * @return {Array}
+   * @return {Set<Mark>}
    */
 
-  getMarksAtCollapsedRangeAsArray(range) {
-    if (range.isUnset) return []
-
-    const { startKey, startOffset } = range
-
-    if (startOffset == 0) {
-      const previous = this.getPreviousText(startKey)
-      if (!previous || previous.text.length == 0) return []
-      if (
-        this.getClosestBlock(startKey) !== this.getClosestBlock(previous.key)
-      ) {
-        return []
+  getActiveMarksAtRange(range) {
+    range = range.normalize(this)
+    if (range.isUnset) return Set()
+    if (range.isCollapsed)
+      return this.getMarksAtPosition(range.startKey, range.startOffset).toSet()
```

Comment:
Can you change these types of long returns to use `{}` instead of being broken across lines like this?

Reply:
Fixed~